rosetta/transactions/parse_transaction.go

```diff
@@ -15,16 +15,208 @@
 package transactions
 
 import (
+	"bytes"
 	"fmt"
+	"strconv"
 
+	"github.com/onflow/cadence/encoding/json"
 	"github.com/onflow/flow-go-sdk"
 
+	"github.com/optakt/flow-dps/models/dps"
+	"github.com/optakt/flow-dps/rosetta/failure"
 	"github.com/optakt/flow-dps/rosetta/identifier"
 	"github.com/optakt/flow-dps/rosetta/object"
 )
 
-// ParseTransactions processes the flow transaction and translates it to a list of operations and a list of
-// signers.
+const (
+	// Transactions should have exactly one authorizer.
+	authorizersRequired = 1
+
+	// Transaction script should have exactly two arguments.
+	argsRequired = 2
+)
+
+// ParseTransactions processes the flow transaction, validates its correctness and translates it
+// to a list of operations and a list of signers.
 func (p *Parser) ParseTransaction(tx *flow.Transaction) ([]object.Operation, []identifier.Account, error) {
-	return nil, nil, fmt.Errorf("TBD: not implemented")
+
+	// Validate the transaction actors. We expect a single authorizer - the sender account.
+	// For now, the sender must also be the proposer and the payer for the transaction.
+
+	if len(tx.Authorizers) != authorizersRequired {
+		return nil, nil, failure.InvalidAuthorizers{
+			Have:        uint(len(tx.Authorizers)),
+			Want:        authorizersRequired,
+			Description: failure.NewDescription("invalid number of authorizers"),
+		}
+	}
+
+	authorizer := tx.Authorizers[0]
+	sender := identifier.Account{
+		Address: authorizer.String(),
+	}
+
+	// Validate the sender address.
+	err := p.validate.Account(sender)
+	if err != nil {
+		return nil, nil, fmt.Errorf("invalid sender account: %w", err)
+	}
+
+	// Verify that the sender is the payer and the proposer.
+	if tx.Payer != authorizer {
+		return nil, nil, failure.InvalidPayer{
+			Have:        tx.Payer,
+			Want:        authorizer,
+			Description: failure.NewDescription("invalid transaction payer"),
+		}
+	}
+	if tx.ProposalKey.Address != authorizer {
+		return nil, nil, failure.InvalidProposer{
+			Have:        tx.ProposalKey.Address,
+			Want:        authorizer,
+			Description: failure.NewDescription("invalid transaction proposer"),
+		}
+	}
+
+	// Verify the transaction script is the token transfer script.
+	script, err := p.generate.TransferTokens(dps.FlowSymbol)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not generate transfer script: %w", err)
+	}
+	if !bytes.Equal(script, tx.Script) {
```

Comment:
Here we're doing an exact match of the transaction text. Since we're the ones generating the script, if it's different we can know it's been touched. I figured that's a freedom we can take.

Do you think that's okay or should I still include some script parsing and come up with something more elaborate? However, after working hard to avoid parsing of Rosetta numbers, parsing Cadence scripts feels even more dangerous..

Reply:
That's an elegant way to do it, actually.